jailbreakd/main.m

```diff
@@ -147,146 +53,100 @@ struct __attribute__((__packed__)) JAILBREAKD_DUMP_CRED {
 
-int runserver(){
-    printf("[jailbreakd] Process Start!\n");
-
-    int32_t rtrn = 0;
-    kern_return_t err;
-    mach_port_t pass_port;
-
-    mach_port_t bootstrap_port = MACH_PORT_NULL;
-    mach_port_t port = MACH_PORT_NULL;
-    tfpzero = MACH_PORT_NULL;
-
-    printf("[jailbreakd] Get Bootstrap Port!\n");
-    
-    /* In the child process grab the port passed by the parent. */
-    err = task_get_special_port(mach_task_self(), SPECIAL_PORT, &bootstrap_port);
-    if(err != KERN_SUCCESS)
-    {
-        mach_error("Can't get bootstrap port:\n", err);
-        return (-1);
-    }
+#define MEMORYSTATUS_CMD_SET_JETSAM_TASK_LIMIT 6
+int memorystatus_control(uint32_t command, int32_t pid, uint32_t flags, void *buffer, size_t buffersize);
 
-    sleep(1);
+int remove_memory_limit(void) {
+    // daemons run under launchd have a very stingy memory limit by default, we need
+    // quite a bit more for patchfinder so disable it here
+    // (note that we need the com.apple.private.memorystatus entitlement to do so)
+    pid_t my_pid = getpid();
+    return memorystatus_control(MEMORYSTATUS_CMD_SET_JETSAM_TASK_LIMIT, my_pid, 0, NULL, 0);
+}
 
-    printf("[jailbreakd] Get Special Port!\n");
-    
-    /* In the child process grab the port passed by the parent. */
-    err = task_get_special_port(mach_task_self(), SPECIAL_PORT, &pass_port);
-    if(err != KERN_SUCCESS)
-    {
-        mach_error("Can't get special port:\n", err);
-        return (-1);
-    }
-    
-    printf("[jailbreakd] Create Mach Port!\n");
-    /* Create a port with a send right. */
-    if(setup_recv_port(&port) != 0)
-    {
-        printf("Can't setup mach port\n");
-        return (-1);
-    }
-    
-    printf("[jailbreakd] Send Mach Port!\n");
-    /* Send port to parent. */
-    rtrn = send_port(pass_port, port);
-    if(rtrn < 0)
-    {
-        printf("Can't send port\n");
-        return (-1);
-    }
-    
-    /* Receive tfp0 from the parent. */
-    rtrn = recv_port(port, &tfpzero);
-    if(rtrn < 0)
-    {
-        printf("Can't receive task port\n");
-        return (-1);
-    }
-    printf("[jailbreakd] Got tfp0: %u\n", tfpzero);
-    
-    /* Set the bootstrap port back to normal. */
-    err = task_set_special_port(mach_task_self(), SPECIAL_PORT, bootstrap_port);
-    if(err != KERN_SUCCESS)
-    {
-        mach_error("Can't set special port:\n", err);
-        return (-1);
+int runserver(){
+    NSLog(@"[jailbreakd] Process Start!\n");
+    remove_memory_limit();
+
+    kern_return_t err = host_get_special_port(mach_host_self(), HOST_LOCAL_NODE, 4, &tfpzero);
+    if (err != KERN_SUCCESS) {
+        NSLog(@"host_get_special_port 4: %s", mach_error_string(err));
+        return 5;
     }
 
     offsets_init();
 
     init_kernel(kernel_base, NULL);
     // Get the slide
     kernel_slide = kernel_base - 0xFFFFFFF007004000;
-    printf("[jailbreakd] slide: 0x%016llx\n", kernel_slide);
+    NSLog(@"[jailbreakd] slide: 0x%016llx\n", kernel_slide);
```

Comment:
do we really need NSLog over printf?

Reply:
Another thing about NSLog: it adds its own \n, so the explicit newline can be removed

Reply:
yeah, that's what i was meaning by that